src/nbl/video/IOpenGL_PhysicalDeviceBase.h

```diff
@@ -561,11 +561,52 @@ class IOpenGL_PhysicalDeviceBase : public IOpenGLPhysicalDeviceBase
 		}
 		else m_glfeatures.MaxAnisotropy = 0u;
 
+		m_features.fullDrawIndexUint32 = false; // TODO
+		m_features.independentBlend = false; // TODO
 		if (m_glfeatures.isFeatureAvailable(m_glfeatures.NBL_ARB_geometry_shader4))
 		{
-			GetIntegerv(GL_MAX_GEOMETRY_OUTPUT_VERTICES, &num);
-			m_glfeatures.MaxGeometryVerticesOut = static_cast<uint32_t>(num);
 			m_features.geometryShader = true;
+			GetIntegerv(GL_MAX_GEOMETRY_SHADER_INVOCATIONS, reinterpret_cast<GLint*>(&m_properties.limits.maxGeometryShaderInvocations));
+			GetIntegerv(GL_MAX_GEOMETRY_INPUT_COMPONENTS, reinterpret_cast<GLint*>(&m_properties.limits.maxGeometryInputComponents));
+			GetIntegerv(GL_MAX_GEOMETRY_OUTPUT_COMPONENTS, reinterpret_cast<GLint*>(&m_properties.limits.maxGeometryOutputComponents));
+			GetIntegerv(GL_MAX_GEOMETRY_OUTPUT_VERTICES, reinterpret_cast<GLint*>(&m_properties.limits.maxGeometryOutputVertices));
+			GetIntegerv(GL_MAX_GEOMETRY_TOTAL_OUTPUT_COMPONENTS, reinterpret_cast<GLint*>(&m_properties.limits.maxGeometryTotalOutputComponents));
+			m_glfeatures.MaxGeometryVerticesOut = m_properties.limits.maxGeometryOutputVertices;
+		}
+		if (m_glfeatures.isFeatureAvailable(m_glfeatures.NBL_ARB_tessellation_shader))
+		{
+			m_features.tessellationShader = true;
+			GetIntegerv(GL_MAX_TESS_GEN_LEVEL, reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationGenerationLevel));
+			// GL_MAX_PATCH_VERTICES => maximum patch size
+			GetIntegerv(GL_MAX_PATCH_VERTICES, reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationPatchSize));
+			// GL_MAX_TESS_CONTROL_INPUT_COMPONENTS => num. components for per-vertex inputs in tess.
+			GetIntegerv(GL_MAX_TESS_CONTROL_INPUT_COMPONENTS, reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationControlPerVertexInputComponents));
+			// GL_MAX_TESS_CONTROL_OUTPUT_COMPONENTS => num. components for per-vertex outputs in tess. 
+			GetIntegerv(GL_MAX_TESS_CONTROL_OUTPUT_COMPONENTS, reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationControlPerVertexOutputComponents)); 
+			// GL_MAX_TESS_PATCH_COMPONENTS => num. components for per-patch output varyings
+			GetIntegerv(GL_MAX_TESS_PATCH_COMPONENTS, reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationControlPerPatchOutputComponents));
+			GetIntegerv(GL_MAX_TESS_CONTROL_TOTAL_OUTPUT_COMPONENTS, reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationControlTotalOutputComponents));
+			GetIntegerv(GL_MAX_TESS_EVALUATION_INPUT_COMPONENTS, reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationEvaluationInputComponents));
+			GetIntegerv(GL_MAX_TESS_EVALUATION_OUTPUT_COMPONENTS, reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationEvaluationOutputComponents));
```

Comment:
careful about GL enums, different extensions have different suffices (EXT,INTEL,OES) and the actual number values of the defines might be different

Reply:
let me introduce the X-Macro
```cpp
#define GLENUM_WITH_SUFFIX(E,S) NBL_EVAL(E##S)
```

you'd make a `src/nbl/video/GL/limit_queries/tessellation_shader.h`
```cpp
		m_features.tessellationShader = true;
			GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_TESS_GEN_LEVEL,GLENUM_SUFFIX), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationGenerationLevel));
			// GL_MAX_PATCH_VERTICES => maximum patch size
			GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_PATCH_VERTICES,GLENUM_SUFFIX), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationPatchSize));
			// GL_MAX_TESS_CONTROL_INPUT_COMPONENTS => num. components for per-vertex inputs in tess.
			GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_TESS_CONTROL_INPUT_COMPONENTS,GLENUM_SUFFIX), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationControlPerVertexInputComponents));
			// GL_MAX_TESS_CONTROL_OUTPUT_COMPONENTS => num. components for per-vertex outputs in tess. 
			GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_TESS_CONTROL_OUTPUT_COMPONENTS,GLENUM_SUFFIX), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationControlPerVertexOutputComponents)); 
			// GL_MAX_TESS_PATCH_COMPONENTS => num. components for per-patch output varyings
			GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_TESS_PATCH_COMPONENTS,GLENUM_SUFFIX), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationControlPerPatchOutputComponents));
			GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_TESS_CONTROL_TOTAL_OUTPUT_COMPONENTS,GLENUM_SUFFIX), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationControlTotalOutputComponents));
			GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_TESS_EVALUATION_INPUT_COMPONENTS,GLENUM_SUFFIX), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationEvaluationInputComponents));
			GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_TESS_EVALUATION_OUTPUT_COMPONENTS,GLENUM_SUFFIX), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationEvaluationOutputComponents));
```

```cpp
if (!isGLES || Version>=MAKE_VERSION(3,2,0))
{
   #define GLENUM_SUFFIX
   #include "src/nbl/video/GL/limit_queries/tessellation_shader.h"
   #undef GLENUM_SUFFIX
}
else if (m_glfeatures.isFeatureAvailable(m_glfeatures.NBL_OES_tessellation_shader))
{
   #define GLENUM_SUFFIX _OES
   #include "src/nbl/video/GL/limit_queries/tessellation_shader.h"
   #undef GLENUM_SUFFIX
}
else if (m_glfeatures.isFeatureAvailable(m_glfeatures.NBL_EXT_tessellation_shader))
{
   #define GLENUM_SUFFIX _OES
   #include "src/nbl/video/GL/limit_queries/tessellation_shader.h"
   #undef GLENUM_SUFFIX
}
else if (m_glfeatures.isFeatureAvailable(m_glfeatures.NBL_INTEL_tessellation_shader))
{
   #define GLENUM_SUFFIX _OES
   #include "src/nbl/video/GL/limit_queries/tessellation_shader.h"
   #undef GLENUM_SUFFIX
}
```

Reply:
1. There are some issues with the macro because S in `GLENUM_WITH_SUFFIX(E,S)` is not evaluated

`GLENUM_WITH_SUFFIX(GL_MAX_TESS_GEN_LEVEL,GLENUM_SUFFIX)`
will be evaluated to 
`GL_MAX_TESS_GEN_LEVEL_GLENUM_SUFFIX` instead of `GL_MAX_TESS_GEN_LEVEL_GLENUM_OES` for examples

But if we do it in two levels then the `GL_MAX_TESS_GEN_LEVEL is going to get evaulated to it's number but we only want it's string.


2. Also `GL_MAX_TESS_GEN_LEVEL` with all it's suffixes have the same value, why are we querying with different suffixes if they are actually the same value `0x8E7D` ?

Reply:
The ONLY solution I found was doing something like this:
```c++
if (!IsGLES || m_glfeatures.Version >= 320u)
{
	#define GLENUM_WITH_SUFFIX(X) X
	#include "nbl/video/GL/limit_queries/tessellation_shader.h"
	#undef GLENUM_WITH_SUFFIX
}
else if (m_glfeatures.isFeatureAvailable(m_glfeatures.NBL_OES_tessellation_shader))
{
	#define GLENUM_WITH_SUFFIX(X) X##_OES
	#include "nbl/video/GL/limit_queries/tessellation_shader.h"
	#undef GLENUM_WITH_SUFFIX
}
```

where the `tessellation_shader.h` would look like:
```c++
m_features.tessellationShader = true;
GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_TESS_GEN_LEVEL), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationGenerationLevel));
.
.
.

```

Reply:
> The ONLY solution I found was doing something like this:
> 
> ```c++
> if (!IsGLES || m_glfeatures.Version >= 320u)
> {
> 	#define GLENUM_WITH_SUFFIX(X) X
> 	#include "nbl/video/GL/limit_queries/tessellation_shader.h"
> 	#undef GLENUM_WITH_SUFFIX
> }
> else if (m_glfeatures.isFeatureAvailable(m_glfeatures.NBL_OES_tessellation_shader))
> {
> 	#define GLENUM_WITH_SUFFIX(X) X##_OES
> 	#include "nbl/video/GL/limit_queries/tessellation_shader.h"
> 	#undef GLENUM_WITH_SUFFIX
> }
> ```
> 
> where the `tessellation_shader.h` would look like:
> 
> ```c++
> m_features.tessellationShader = true;
> GetIntegerv(GLENUM_WITH_SUFFIX(GL_MAX_TESS_GEN_LEVEL), reinterpret_cast<GLint*>(&m_properties.limits.maxTessellationGenerationLevel));
> .
> .
> .
> ```

decent enough, I approve